core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -613,6 +618,12 @@ final class ShuffleBlockFetcherIterator(
           }
           if (isNetworkReqDone) {
             reqsInFlight -= 1
+            if (!buf.isInstanceOf[NettyManagedBuffer]) {
+              // Non-`NettyManagedBuffer` doesn't occupy Netty's memory so we can unset the flag
+              // directly once the request succeeds. But for the `NettyManagedBuffer`, we'll only
+              // unset the flag when the data is fully consumed (see `BufferReleasingInputStream`).
+              NettyUtils.isNettyOOMOnShuffle = false
```

Comment:
This flag is in `NettyUtils` so I am assuming that if there are multiple tasks with their own iterators, then they all will be setting/unsetting the same flag. Is that correct? 
Wouldn't this decision to unset the flag here, when it is a non-netty_managed_buffer, by a single iterator cause a problem?

Reply:
Good point!

After a second thinking, I think the condition should be " if (isRemoteRequest && !buf.isInstanceOf[NettyManagedBuffer])". WDYT?

Reply:
Oh wait...`isNetworkReqDone` already indicates it's a remote request. And a remote request would always consume Netty's memory. So, I think the current condition should be fine?

Reply:
> This flag is in NettyUtils so I am assuming that if there are multiple tasks with their own iterators, then they all will be setting/unsetting the same flag. Is that correct?

Correct

Reply:
My concern here is that in this case, an iterator checks if it is not a `NettyManagedBuffer` and then doesn't wait to release the buffer and unsets the flag. But what if another instance which has set `NettyUtils.isNettyOOMOnShuffle=true` because it saw netty OOM and it can recover only when nettyManagedBuffers are released?

Reply:
How about this:

```scala
if (Netty.freeMemory > minRequestSize) {
  unset the flag
}
```
?
* `minRequestSize` supposed to be the minimize request size of all the pending requests(or block?).

I think this would mitigate the situation you mentioned.

Reply:
FYI: to simplify the implementation, I used the average block size here at the end.

Reply:
So the `avgBlockSize` is specific for a particular iterator instance. Just thinking if unsetting the flag on this information will create a problem for other instances...

Reply:
So assume the avgA is for the iterator A and avgB is for the iterator B and we're going to unset the flag when freeMemory > avgA:

If avgA > avgB,  I think this shouldn't be a problem for iterator B.

If avgA < avgB, although flag is unset, but please note that iterator B won't start the next request until an in-flight request finished, which would release the avgB memory. So I think it should also be fine. 



BTW, it could be complex if we want to maintain a global min average. For example, if the current global min average is avgA and iterator A finished, we need to update the global min average, which can be complex to compute at this time.